src/EventListener/UserSessionListener.php

```diff
@@ -0,0 +1,274 @@
+<?php
+
+/**
+ * This file is part of Contao.
+ *
+ * Copyright (c) 2005-2015 Leo Feyer
+ *
+ * @license LGPL-3.0+
+ */
+
+namespace Contao\CoreBundle\EventListener;
+
+use Contao\BackendUser;
+use Contao\CoreBundle\Session\Attribute\AttributeBagAdapter;
+use Contao\FrontendUser;
+use Doctrine\DBAL\Driver\Connection;
+use Symfony\Component\HttpFoundation\Request;
+use Symfony\Component\HttpFoundation\Session\Attribute\AttributeBagInterface;
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\HttpKernel\Event\FilterResponseEvent;
+use Symfony\Component\HttpKernel\Event\GetResponseEvent;
+
+/**
+ * Stores the session data of back end and front end users and restores it on
+ * the next login so the last application state can be restored.
+ *
+ * @author Yanick Witschi <https://github.com/toflar>
+ */
+class UserSessionListener extends ScopeAwareListener
+{
+    /**
+     * @var SessionInterface
+     */
+    private $session;
+
+    /**
+     * @var Connection
+     */
+    private $connection;
+
+    /**
+     * Constructor.
+     *
+     * @param SessionInterface $session
+     */
+    public function __construct(
+        SessionInterface $session,
+        Connection $connection
+    ) {
+        $this->session = $session;
+        $this->connection = $connection;
+    }
+
+    /**
+     * Checks whether the user has session data stored and passes it to the
+     * session service if so.
+     *
+     * @param GetResponseEvent $event The event object
+     */
+    public function onKernelRequest(GetResponseEvent $event)
+    {
+        if (!$this->isFrontendMasterRequest($event)
```

Comment:
Yes, which is correct in my opinion. If you want to have session data you should take the session of the master request.